sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeExpression.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.types
+
+import org.apache.spark.sql.catalyst.expressions.Expression
+
+abstract class DataTypeExpression(val dataType: DataType) {
+  /**
+   * Enables matching against DataType for expressions:
+   * {{{
+   *   case Cast(child @ BinaryType(), StringType) =>
+   *     ...
+   * }}}
+   */
+  private[sql] def unapply(e: Expression): Boolean = e.dataType == dataType
+}
+
+abstract class AtomicTypeExpression(override val dataType: DataType)
+  extends DataTypeExpression(dataType)
+
+abstract class DatetimeTypeExpression(override val dataType: DataType)
```

Comment:
Seems the above two abstract classes are not needed.

Reply:
oh yes good point!